File: python/avin/model/instrument_kind.py
# ────────────────────────────────────────────────────────────────────────────
#  AVIN
#  Understand the market before trading it.
#
#  https://avin.info
# ────────────────────────────────────────────────────────────────────────────

from __future__ import annotations

from enum import Enum

from avin._native import PyInstrumentKind


class InstrumentKind(Enum):
    """
    Financial instrument kind.

    Examples
    --------
    >>> for kind in InstrumentKind:
    ...     print(kind)
    Currency
    Index
    Stock
    Future
    Bond
    Option
    ETF
    >>> kind = InstrumentKind.FUTURE
    >>> str(kind)
    'Future'
    >>> InstrumentKind.from_str("bond") == InstrumentKind.STOCK
    True
    >>> InstrumentKind.from_str("stock") is InstrumentKind.STOCK
    True
    """

    CURRENCY = PyInstrumentKind.Currency
    INDEX = PyInstrumentKind.Index
    STOCK = PyInstrumentKind.Stock
    FUTURE = PyInstrumentKind.Future
    BOND = PyInstrumentKind.Bond
    OPTION = PyInstrumentKind.Option
    ETF = PyInstrumentKind.ETF

    _inner: PyInstrumentKind

    def __new__(cls, inner: PyInstrumentKind):
        obj = object.__new__(cls)
        obj._value_ = inner.display()
        obj._inner = inner

        return obj

    def __str__(self) -> str:
        return self._inner.display()

    @classmethod
    def from_str(cls, s: str) -> InstrumentKind:
        """
        Parse an instrument kind.

        Parsing is case-insensitive.

        Parameters
        ----------
        s : str
            Instrument kind.

        Raises
        ------
        ValueError
            If the instrument kind is unknown.
        RuntimeError
            If the native and public Python instrument kind definitions are
            out of sync.

        Examples
        --------
        >>> InstrumentKind.from_str("FuTuRe") is InstrumentKind.FUTURE
        True
        >>> InstrumentKind.from_str("ETF") is InstrumentKind.ETF
        True
        """
        return InstrumentKind._from_native(PyInstrumentKind.from_str(s))

    @classmethod
    def _from_native(cls, inner: PyInstrumentKind) -> InstrumentKind:
        for kind in cls:
            if kind._inner.eq(inner):
                return kind

        raise RuntimeError(
            "native instrument kind is missing from InstrumentKind"
        )

File: crates/avin_model/src/instrument_kind.rs
// ───────────────────────────────────────────────────────────────────────────
// AVIN
// Understand the market before trading it.
//
// https://avin.info
// ───────────────────────────────────────────────────────────────────────────

use std::fmt::Display;
use std::str::FromStr;

use avin_utils::AvinError;

#[derive(Debug, Clone, Copy, PartialEq, Eq)]
pub enum InstrumentKind {
    Currency,
    Index,
    Stock,
    Future,
    Bond,
    Option,
    ETF,
}

/// Instrument kind.
///
/// Represents the kind of a financial instrument.
///
/// # Examples
///
/// ```
/// use std::str::FromStr;
///
/// use avin_model::InstrumentKind;
///
/// for kind in InstrumentKind::all() {
///     println!("{kind}");
/// }
///
/// // Parsing is case-insensitive.
/// let kind = InstrumentKind::from_str("future").unwrap();
/// assert_eq!(kind, InstrumentKind::Future);
/// ```
impl InstrumentKind {
    /// Returns all supported instrument kinds.
    pub const fn all() -> &'static [Self] {
        &[
            Self::Currency,
            Self::Index,
            Self::Stock,
            Self::Future,
            Self::Bond,
            Self::Option,
            Self::ETF,
        ]
    }
}

impl Display for InstrumentKind {
    fn fmt(&self, f: &mut std::fmt::Formatter<'_>) -> std::fmt::Result {
        match self {
            Self::Currency => f.write_str("Currency"),
            Self::Index => f.write_str("Index"),
            Self::Stock => f.write_str("Stock"),
            Self::Future => f.write_str("Future"),
            Self::Bond => f.write_str("Bond"),
            Self::Option => f.write_str("Option"),
            Self::ETF => f.write_str("ETF"),
        }
    }
}

impl FromStr for InstrumentKind {
    type Err = AvinError;

    /// Parses an instrument kind.
    ///
    /// Parsing is case-insensitive.
    ///
    /// # Errors
    ///
    /// Returns an error if the instrument kind is unknown.
    ///
    /// # Examples
    ///
    /// ```
    /// use std::str::FromStr;
    ///
    /// use avin_model::InstrumentKind;
    ///
    /// assert_eq!(
    ///     InstrumentKind::from_str("FuTuRe").unwrap(),
    ///     InstrumentKind::Future
    /// );
    /// assert_eq!(
    ///     InstrumentKind::from_str("ETF").unwrap(),
    ///     InstrumentKind::ETF
    /// );
    ///
    /// assert!(InstrumentKind::from_str("foo").is_err());
    /// ```
    fn from_str(s: &str) -> Result<Self, Self::Err> {
        match s.to_ascii_lowercase().as_str() {
            "currency" => Ok(Self::Currency),
            "index" => Ok(Self::Index),
            "stock" => Ok(Self::Stock),
            "future" => Ok(Self::Future),
            "bond" => Ok(Self::Bond),
            "option" => Ok(Self::Option),
            "etf" => Ok(Self::ETF),
            _ => {
                let available = Self::all()
                    .iter()
                    .map(Self::to_string)
                    .collect::<Vec<_>>()
                    .join(", ");

                let msg = format!(
                    "unknown instrument kind '{}', available=[{}]",
                    s, available
                );

                Err(AvinError::InvalidValue(msg))
            }
        }
    }
}

#[cfg(test)]
mod tests {
    use super::*;

    #[test]
    fn all() {
        let expected = [
            InstrumentKind::Currency,
            InstrumentKind::Index,
            InstrumentKind::Stock,
            InstrumentKind::Future,
            InstrumentKind::Bond,
            InstrumentKind::Option,
            InstrumentKind::ETF,
        ];

        assert_eq!(InstrumentKind::all(), expected);
    }

    #[test]
    fn display() {
        assert_eq!(InstrumentKind::Currency.to_string(), "Currency");
        assert_eq!(InstrumentKind::Index.to_string(), "Index");
        assert_eq!(InstrumentKind::Stock.to_string(), "Stock");
        assert_eq!(InstrumentKind::Future.to_string(), "Future");
        assert_eq!(InstrumentKind::Bond.to_string(), "Bond");
        assert_eq!(InstrumentKind::Option.to_string(), "Option");
        assert_eq!(InstrumentKind::ETF.to_string(), "ETF");
    }

    #[test]
    fn from_str() {
        assert_eq!(
            InstrumentKind::from_str("Currency").unwrap(),
            InstrumentKind::Currency
        );
        assert_eq!(
            InstrumentKind::from_str("Index").unwrap(),
            InstrumentKind::Index
        );
        assert_eq!(
            InstrumentKind::from_str("Stock").unwrap(),
            InstrumentKind::Stock
        );
        assert_eq!(
            InstrumentKind::from_str("Future").unwrap(),
            InstrumentKind::Future
        );
        assert_eq!(
            InstrumentKind::from_str("BoNd").unwrap(),
            InstrumentKind::Bond
        );
        assert_eq!(
            InstrumentKind::from_str("OPTION").unwrap(),
            InstrumentKind::Option
        );
        assert_eq!(
            InstrumentKind::from_str("etf").unwrap(),
            InstrumentKind::ETF
        );

        assert!(InstrumentKind::from_str("foo").is_err());
    }
}
